### PR TITLE
Add lazy_bench.py to measure trace overhead and compute efficiency

### PR DESCRIPTION
**Pull request by wconstab**
This tool has 2 new 'experiments' and builds on infra from @jansel's torchdynamo bench script [torchdynamo/torchbench.py](https://github.com/jansel/torchdynamo/blob/main/torchbench.py ).

The infra components take care of 
* iterating over torchbench models in a more convenient way, handling filtering, errors
* correctness checks
* mixing in non-torchbenchmark benchmarks
* interleaving measurements of the control/experiment and computing statistical significance
* hooks for synchronization that can be specialized for cuda or lazytensor
* custom sync modes to allow syncing after every step or running many async steps before syncing

The overhead experiment compares the lazy trace overhead with the full cuda execution time.  This may sound like a strange choice, but the point is to provide a reference point of how much time lazy tracing takes with reference to the time eager normally spends in execution.  The expectation is a small fraction.
* an alternative approach is to measure lazy tracing against the portion of eager that launches cuda kernels but never sync.  If we could guarantee that some of the time, the cuda driver didn't force syncs, this would be fair.  It wouldn't work for CPU.  At least the full-sync way is consistent
* another alternative is to compare lazy trace overhead to execution with meta-tensors.  We don't expect meta-tensors to work for this case, since we know that many of the ops we lazy-trace are not yet structured kernels.
image

The compute experiment performs synchronization and compares the full cost of execution of an eager device vs a lazy device.  There are two flavors- Unamortized and Amortized.  Unamortized means you synchronize every step, which is unfavorable for both lazy and cuda, (but fine for cpu).  Amortized means you run N steps asynchronously before finally synchronizing.  Each measurement includes at least one begin and one final sync.

image

Notes/Todos/Opens
* run the same experiments using cpu, flush out any issues
* Ideally the LazySync and CudaSync classes would be sufficient; The ToDeviceSync class is added as a sanity check, and indeed disagrees with LazySync implying something isn't quite right yet
* debug the deltas between different sync modes for cuda/lazycuda- expect more similar results.

### CODE REVIEW

File: lazy_tensor_core/lazy_bench.py [hunk 1]
```diff
@@ -0,0 +1,358 @@
...
+# from caffe2.python import workspace
+# workspace.GlobalInit(['caffe2', '--caffe2_log_level=-5'])
+
+import lazy_tensor_core
+import lazy_tensor_core.core.lazy_model as ltm
+import lazy_tensor_core.debug.metrics as met
+lazy_tensor_core._LAZYC._ltc_init_ts_backend()
+
+os.environ["KALDI_ROOT"] = "/tmp"  # avoids some spam
+torchbench_dir = abspath("/home/whc/benchmark")
```
Comment (jansel) [inline lazy_tensor_core/lazy_bench.py:38]: Should update this to something dynamic before landing

**Comment by jansel** [comment]
Can you share the results?

**Comment by wconstab** [comment]
I fixed the worst of issues with the timing script - hf_bart timings were way different between LazySync and ToDeviceSync.  The bug was in the `to_device` function where I specially handle output classes from hf models.  My handling was broken, and for Bart, that was catastrophic since Bart returns all of its model hidden states as attributes of its output object, and to_device was recursively copying them all out.  I now only copy the logits out in 'to_device'.  Bart timings are consistent now.

**Review by desertfire — APPROVED** [review]
Stamp to move fast

**Comment by wconstab** [comment]
This tool has 2 new 'experiments' and builds on infra from @jansel's torchdynamo bench script torchdynamo/torchbench.py.

The infra components take care of

* iterating over torchbench models in a more convenient way, handling filtering, errors
* correctness checks
* mixing in non-torchbenchmark benchmarks
* interleaving measurements of the control/experiment and computing statistical significance
* hooks for synchronization that can be specialized for cuda or lazytensor
* custom sync modes to allow syncing after every step or running many async steps before syncing

The overhead experiment compares the lazy trace overhead with the full cuda execution time. This may sound like a strange choice, but the point is to provide a reference point of how much time lazy tracing takes with reference to the time eager normally spends in execution. The expectation is a small fraction.

an alternative approach is to measure lazy tracing against the portion of eager that launches cuda kernels but never sync. If we could guarantee that some of the time, the cuda driver didn't force syncs, this would be fair. It wouldn't work for CPU. At least the full-sync way is consistent
another alternative is to compare lazy trace overhead to execution with meta-tensors. We don't expect meta-tensors to work for this case, since we know that many of the ops we lazy-trace are not yet structured kernels.
